src/Database/LSMTree/Internal.hs

```diff
@@ -680,6 +684,65 @@ updates resolve es th = do
             es
             tc
 
+{-------------------------------------------------------------------------------
+  Blobs
+-------------------------------------------------------------------------------}
+
+retrieveBlobs ::
+     m ~ IO  -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => Session m h
+  -> V.Vector (BlobRef (Run m (FS.Handle h)))
+  -> m (V.Vector SerialisedBlob)
+retrieveBlobs sesh refs =
+    withOpenSession sesh $ \seshEnv ->
+      bracket_ upgradeWeakReferences removeReferences $ do
+
+        -- Prepare the IOOps:
+        -- We use a single large memory buffer, with appropriate offsets within
+        -- the buffer.
+        let bufSize :: Int
+            !bufSize = V.sum (V.map blobRefSpanSize refs)
+
+            {-# INLINE bufOffs #-}
+            bufOffs :: V.Vector Int
+            bufOffs = V.scanl (+) 0 (V.map blobRefSpanSize refs)
+        buf <- P.newPinnedByteArray bufSize
+        let ioops = V.zipWith (Run.readBlobIOOp buf) bufOffs refs
+            hbio  = sessionHasBlockIO seshEnv
+
+        -- Submit the IOOps all in one go:
+        _ <- FS.submitIO hbio ioops
+        -- We do not need to inspect the results because IO errors are
+        -- thrown as exceptions, and the result is just the read length
+        -- which is already known. Short reads can't happen here.
```

Comment:
We should probably change [this code](https://github.com/well-typed/blockio-uring/blob/main/src/System/IO/BlockIO/URing.hs#L111-L140). Do the pattern synonyms do anything? If we can always expect `IOResult` to contain a non-negative integer?

We are currently explicitly handling the case where an `IOResult` can contain an error number in https://github.com/IntersectMBO/lsm-tree/blob/main/blockio-api/src-linux/System/FS/BlockIO/Async.hs#L75-L99. It should probably be a catch+rethrow instead

Reply:
The `blockio-uring` lib does not throw exceptions. The `IOError` and `IOResult` patters do what you'd expect.

It's our wrapper in `blockio-api` that changes the `IOResult` to have a single constructor for the successful case. And so the instance for `blockio-uring` does that by throwing exceptions.

The code above is using our wrapper, and the comment refers to that, not to the underlying `blockio-uring`.